Fix custom scheme handling in wxWebView WebKit2 implementation

swt2c opened the pull request:
The custom scheme handling implementation had been inherited from the original
WebKit1 implementation.  It attempted to intercept navigation and resource load
requests and then inject the resources.  It seems that this method doesn't work
in WebKit2, but fortunately, there is native support in WebKit2 for custom URI
schemes through the webkit_web_context_register_uri_scheme() API.

swt2c commented:
Please cherry-pick this to WX_3_0_BRANCH also.  If it doesn't apply, let me know.  I'll do the backport.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -325,49 +289,55 @@ wxgtk_webview_webkit_title_changed(GtkWidget* widget,
+        }
+        else
+        {
+            GError *error = g_error_new(WEBKIT_NETWORK_ERROR,
```
It would be worth extending `wxGtkError` class from `include/wx/gtk/private/error.h` to have a non-default ctor and use it here to avoid manual calls to `g_error_free()`.

swt2c replied:
Done.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -569,38 +539,35 @@ bool wxWebViewWebKit::Create(wxWindow *parent,
-                           "initialize-web-extensions",
-                           G_CALLBACK(wxgtk_initialize_web_extensions),
-                           m_dbusServer);
+    g_signal_connect(webkit_web_context_get_default(),
```
Just for my education: why were all these `connect_after` changed to simple `connect`s? Were they simply wrong before and we should have never let the default behaviour take place before our callsbacks were invoked?

swt2c replied:
The one for `decide-policy` was definitely wrong, as the default signal handler was getting called before us, and our handler never got called.  As best as I could tell, none of the other ones were definitely breaking anything, but I decided to change them all just in case.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -325,49 +289,55 @@ wxgtk_webview_webkit_title_changed(GtkWidget* widget,
-        //loaded it from the archive
-        if(webKitCtrl->m_vfsurl == uri)
-            return;
+        wxString uri = webkit_uri_scheme_request_get_uri(request);
```
This doesn't look right, it should use `wxString::FromUTF8()` here.

This variable could also be `const`.

swt2c replied:
Updated.

vadz reviewed (changes requested):
Thanks for fixing this!

I'd like to merge it, but `FromUTF8()` really needs to be added.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -325,49 +289,53 @@ wxgtk_webview_webkit_title_changed(GtkWidget* widget,
-            wxString path = "data:" + mime + ";base64," + data;
-            //Then we can redirect the call
-            webkit_uri_request_set_uri(request, path.utf8_str());
+            webkit_uri_scheme_request_finish(request, stream, length, mime);
```
There is also missing `.utf8_str()` here, please see [this post](https://groups.google.com/d/msg/wx-users/XhCdZbPrDAM/gFCAgN4DBwAJ)

swt2c replied:
I did see the post, just not until after I had already pushed.  :(  Fixed.

vadz commented on include/wx/gtk/private/error.h:
```diff
@@ -21,6 +21,7 @@ class wxGtkError
 {
 public:
     wxGtkError() { m_error = NULL; }
+    wxGtkError(GError* error) { m_error = error; }
```
Please make it `explicit`, TIA.

swt2c replied:
Done.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -1279,6 +1244,10 @@ bool wxWebViewWebKit::RunScript(const wxString& javascript, wxString* output)
 {
     m_handlerList.push_back(handler);
+    WebKitWebContext* context = webkit_web_context_get_default();
+    webkit_web_context_register_uri_scheme(context, handler->GetName(),
```
Missing `utf8_str()` here too, see the same post for the error details.

swt2c replied:
Fixed.

vadz commented on src/gtk/webview_webkit2.cpp:
```diff
@@ -325,49 +289,53 @@ wxgtk_webview_webkit_title_changed(GtkWidget* widget,
+        {
+            wxGtkError error(g_error_new(WEBKIT_NETWORK_ERROR,
+                                         WEBKIT_NETWORK_ERROR_FILE_DOES_NOT_EXIST,
+                                         "File not found: %s", uri.mb_str().data()));
```
This `mb_str()` should actually be `utf8_str()` too (also below).

swt2c replied:
Fixed.

vadz reviewed (changes requested):
Sorry, I thought you saw the mailing post, so I didn't update this review, but conversions to `char*` need to be explicit, they're not available in `wxUSE_STL=1` build.

swt2c committed:
Fix custom scheme handling in wxWebView WebKit2 implementation

The custom scheme handling implementation had been inherited from the original
WebKit1 implementation.  It attempted to intercept navigation and resource load
requests and then inject the resources.  It seems that this method doesn't work
in WebKit2, but fortunately, there is native support in WebKit2 for custom URI
schemes through the webkit_web_context_register_uri_scheme() API.

vadz commented:
Thanks again for these changes! Committed to both branches now, but I've only tested compilation, not run-time functionality, in 3.0, please retest if it really works there. TIA!

swt2c commented:
Confirmed, works fine in 3.0 branch also.
